### PR TITLE
Complete rewrite using generic branch and prune with contractors

### PR DESCRIPTION
**Pull request by dpsanders**

### CODE REVIEW

File: src/branch_and_prune.jl [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        NX =  refine(X -> C.op(C.f, C.fp, X), NX)
+        return :unique, NX
+    end
+
+
+    return :unknown, NX
+end
+
+
+roots{C<:Contractor}(f, X, contractor::Type{C}, tol::Float64=1e-3) = branch_and_prune(f, X, contractor, tol)
```
Comment (lbenet) [inline src/branch_and_prune.jl]: Add docstrings for `roots`

**Comment by lbenet** [comment]
I have done some work here for tests to pass. Essentially some functions are currently unexported, among other minor things. Shall I commit directly to this branch, or shall I do it in an alternative one?

**Comment by dpsanders** [comment]
Just add it to this huge branch I think. Thanks!

**Comment by lbenet** [comment]
Just pushed a PR which allows to pass tests locally. 

If you can take a look on the changes it would be worth doing it, because I think your plan was to deprecate/erase some of the previous functions (e.g., use `roots` instead of `newton` or `krawczyk`).

**Comment by lbenet** [comment]
Done; tests are passing...

**Comment by dpsanders** [comment]
Thanks for the tests! I have added a basic docstring for roots.

I think we should merge this and refine later!

**Comment by dpsanders** [comment]
I converted some of examples.jl into new tests, which pass. 
Merging this for now!

**Pull request merged**